File: 백준/C++/단계별/21단계 - 큐, 덱/10866번 - 덱.cpp
#include <iostream>
#include <queue>
#include <string>
using namespace std;
#define MAX 10000

int dq[MAX * 2 + 1];
int frontIdx = MAX, backIdx = MAX;
int n;
string s;
bool empty() {
	if (frontIdx == backIdx) 
		return 1;
	else
		return 0;
}

void push_front(int value) {
	dq[--frontIdx] = value;
}

void push_back(int value) {
	dq[backIdx++] = value;
}

int pop_front() {
	if (empty()) {
		return -1;
	}
	else {
		return dq[frontIdx++];
	}
}

int pop_back() {
	if (empty()) {
		return -1;
	}
	else {
		return dq[--backIdx];
	}
}

int size() {
	return backIdx - frontIdx;
}

int front() {
	if (empty()) {
		return -1;
	}
	else {
		return dq[frontIdx];
	}
}

int back() {
	if (empty()) {
		return -1;
	}
	else {
		return dq[backIdx - 1];
	}
}

int main() {
	ios::sync_with_stdio(false);
	cin.tie(NULL);
	cout.tie(NULL);

	cin >> n;

	while (n--) {
		cin >> s;

		if (s == "push_back") {
			int input;
			cin >> input;
			push_back(input);
		}
		if (s == "push_front") {
			int input;
			cin >> input;
			push_front(input);
		}
		if (s == "pop_back") {
			cout << pop_back() << "\n";
		}
		if (s == "pop_front") {
			cout << pop_front() << "\n";
		}
		if (s == "empty") {
			cout << empty() << "\n";
		}
		if (s == "size") {
			cout << size() << "\n";
		}
		if (s == "front") {
			cout << front() << "\n";
		}
		if (s == "back") {
			cout << back() << "\n";
		}
	}

	return 0;
}	

File: 백준/Python/Silver/10866번 - 덱.py
import sys
from collections import deque

n = int(input())
q = deque()

for _ in range(n):
    command = sys.stdin.readline().split()

    if command[0] == 'push_front':
        q.appendleft(command[1])
    
    elif command[0] == 'push_back':
        q.append(command[1])

    elif command[0] == 'pop_front':
        if q:
            print(q[0])
            del q[0]
        else:
            print(-1)
    
    elif command[0] == 'pop_back':
        if q:
            print(q[-1])
            del q[-1]
        else:
            print(-1)
    
    elif command[0] == 'size':
        print(len(q))

    elif command[0] == 'empty':
        if q:
            print(0)
        else:
            print(1)
    
    elif command[0] == 'front':
        if q:
            print(q[0])
        else:
            print(-1)
    
    elif command[0] == 'back':
        if q:
            print(q[-1])
        else:
            print(-1)